underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java

```diff
@@ -105,56 +96,59 @@ public int read(byte[] b) throws IOException {
 
   @Override
   public int read(byte[] b, int offset, int length) throws IOException {
-    if (length == 0) {
-      return 0;
+    S3Object object = getObject(mPos, mPos + length);
+    if (object == null) {
+      // range request cannot meet
+      object = getObject(mPos);
     }
-    if (mIn == null) {
-      openStream();
+    if (object == null) {
+      // mPos already pass file end
+      return -1;
     }
-    int read = mIn.read(b, offset, length);
-    if (read != -1) {
-      mPos += read;
+    try (S3ObjectInputStream in = object.getObjectContent()) {
+      int ret = in.read(b, offset, length);
+      if (ret == -1) {
+        return ret;
+      }
+      mPos += ret;
+      return ret;
     }
-    return read;
   }
 
-  @Override
-  public long skip(long n) throws IOException {
-    if (n <= 0) {
-      return 0;
-    }
-    closeStream();
-    mPos += n;
-    openStream();
-    return n;
+  private S3Object getObject(long start) {
+    return getObject(start, Long.MAX_VALUE - 1);
   }
 
-  /**
-   * Opens a new stream at mPos if the wrapped stream mIn is null.
-   */
-  private void openStream() throws IOException {
-    if (mIn != null) { // stream is already open
-      return;
-    }
-    GetObjectRequest getReq = new GetObjectRequest(mBucketName, mKey);
+  private S3Object getObject(long start, long end) {
     // If the position is 0, setting range is redundant and causes an error if the file is 0 length
-    if (mPos > 0) {
-      getReq.setRange(mPos);
+    if (start > 0 || end > 0) {
```

Comment:
if start=0 and end =1, this would error out for file with 0 length?

Reply:
will add unit test to make sure it works